### ISSUE TITLE
Lots of warnings during compliation of JTE generated Kotlin code

### ISSUE THREAD

**Issue by IvanPizhenko** [casid/jte#179]
After we have swithed to `generate()` in the `build.gradle.kts` (see our another issue #177), we are getting lots of Kotlin compiler warnings which we have never had before. 

Below is list of warnings that we've faced:

```kotlin
// Unchecked cast: Any? to Response<ServiceResponse>
// many similar ones for different parameter names and types
// these all are our custom parameters which we pass into a KTE page, by declaring them like
// @param keyHistoryDetailsResponse: Response<KeyHistoryDetailsResponse>
    val fesServiceResponse = params["fesServiceResponse"] as Response<com.flowcrypt.eap.clients.fes.ServiceResponse>

// Parameter 'params' is never used
	@JvmStatic fun renderMap(jteOutput:gg.jte.html.HtmlTemplateOutput, jteHtmlInterceptor:gg.jte.html.HtmlInterceptor?, params:Map<String, Any?>) {
		render(jteOutput, jteHtmlInterceptor);
	}

// Parameter 'jteHtmlInterceptor' is never used
	@JvmStatic fun render(jteOutput:gg.jte.html.HtmlTemplateOutput, jteHtmlInterceptor:gg.jte.html.HtmlInterceptor?) {
		jteOutput.writeContent("....")
        }

// 'capitalize(): String' is deprecated. Use replaceFirstChar instead
       jteOutput.writeUserContent(provider.capitalize())
```
I suggest to add proper `@Suppress(...)` statements into the generated code.
In the last case I'd suggest to use recommended method instead of deprecated one.
On my end, I'm looking into how to disable such warnings for certain paths (generated code) in the project but no luck so far.

**Comment by casid** [comment]
Thanks for reporting, I believe those warnings are becoming apparent now, because with `generate()` the Kotlin classes are generated and then simply compiled with the other source files of the project. Which then leads to warnings in the project compilation phase.

I think the proper fix here would be to suppress unused/unchecked warnings in the generated classes.

The last warning on `capitalize()` should probably be fixed in your template, there should be somewhere a `${provider.capitalize()}` that needs to be adjusted to the new method.

**Comment by IvanPizhenko** [comment]
Thanks, found `provider.capitalize()` in my code, will fix it.

**Comment by casid** [comment]
Nice! I just pushed a commit that does a `@Suppress("UNCHECKED_CAST", "UNUSED_PARAMETER")` for all generated Kotlin template classes. After this change I could run a Gradle build with Kotlin templates that don't use `jteHtmlInterceptor` and have generics in their parameters without compiler warnings.

In case you want to verify it on your end, you could use the latest snapshot version.

**Comment by IvanPizhenko** [comment]
Can you please also add in the beginning of each generated file following line:
```kotlin
@file:Suppress("ktlint")
```
I may later provide more fine-grained list of ktlint rules to disable, but for now suppressing all of them will be good enough.

**Comment by casid** [comment]
Ah, good catch. Done!

**Comment by casid** [comment]
I think it should be fine to just suppress the linter for these generated files. They should not be edited by humans.

**Comment by IvanPizhenko** [comment]
Here's more fine-gained list of rules that I had to suppress to make ktlint happy with all generated files for our project:
```kotlin
@file:Suppress(
  "ktlint:indent",
  "ktlint:max-line-length",
  "ktlint:comma-spacing",
  "ktlint:annotation",
  "ktlint:spacing-between-declarations-with-annotations",
  "ktlint:parameter-list-wrapping",
  "ktlint:colon-spacing",
  "ktlint:argument-list-wrapping",
  "ktlint:no-semi",
  "ktlint:import-ordering",
  "ktlint:no-unused-imports",
  "ktlint:no-trailing-spaces"
)
```
I think "suppress all at once" will work just fine, but if you eventually will want to generate not only correct code but also more beautiful and readable, this list will let you know what to optimize in your code generation.

**Comment by casid** [comment]
Thanks! However, that's not on my list atm, since the code generation is optimized for speed rather than beauty :-) But I think this is okay, only in very rare cases a human will look at those files.

**Comment by IvanPizhenko** [comment]
> Ah, good catch. Done!

That's great. Please publish new JTE version with all those fixes and we will try it.

**Comment by IvanPizhenko** [comment]
btw, is this change only in JTE library itself or we will also need to update Gradle plugin?

**Comment by casid** [comment]
This is only a change in `jte-kotlin` module. However, I always keep the same version of all modules when doing a new release, just for simplicity :-) I'll probably do the next release somewhere start of next week.

It would be nice if you could give it a try to ensure everything is working as expected. If you like, you could do that using JitPack: https://jitpack.io/#casid/jte/df5433b923

After temporarily adding the jitpack repo, to swap for the adjusted `jte-kotlin`, you could use this dependency:

```
dependencies {
	implementation 'com.github.casid.jte:jte-kotlin:df5433b923'
}
```

**Comment by IvanPizhenko** [comment]
No, this Jitpack didn't work for me. Maybe I've missed something.
What I did:
1. Added repo itself:
```kotlin
  maven { url = uri("https://jitpack.io") }
```
2. Commented out references to release JTE and added reference to JTE from Jitpack
```kotlin
  api("com.github.casid.jte:jte-kotlin:df5433b923")
  // api("gg.jte:jte-kotlin:2.2.1") // this version must match jte plugin version in root build.gradle.kts
```
3. Removed all generated files.
4. Run build
As result: files are generated in the old way and warnings are here. Looks like it continues to use current release JTE version.
Can't explain myself why this happens. any ideas?
Note, I still have Gradle plugin of version `2.2.1` in my `build.gradle.kts`.

**Comment by casid** [comment]
Ah, I‘m sorry you are right this cannot work, since the jte-kotlin version bundled in the plugin is still old. I‘ll release a new version tomorrow, including the Kotlin downgrade!

**Comment by IvanPizhenko** [comment]
Thank you!

**Comment by casid** [comment]
I just published jte 2.2.2 :-)

**Comment by IvanPizhenko** [comment]
Warnings have gone with JTE 2.2.2, so closing this issue.
